src/Couchbase.Lite.Shared/Query/QueryBase.cs

```diff
@@ -89,108 +117,76 @@ public QueryBase()
         #endregion
 
         #region Public Methods
+
         public void Stop()
         {
-            if (_updating) {
-                _stopping = true;
-            } else {
-                Stopped();
+            Database?.RemoveActiveStoppable(this);
+            foreach (var querier in _liveQueriers) {
+                if (querier != null)
+                    querier.StopObserver();
             }
+
+            _liveQueriers.Clear();
         }
+
         #endregion
 
-        #region Private Methods
+        #region Internal Methods
 
-        private void OnDatabaseChanged(object sender, DatabaseChangedEventArgs e)
+        internal unsafe void CreateQuery()
         {
-            if (_willUpdate)
-            {
-                return;
-            }
-
-            // External updates should poll less frequently
-            var updateInterval = _updateInterval;
+            if (_c4Query == null) {
+                C4Query* query = (C4Query*)ThreadSafety.DoLockedBridge(err =>
+                {
+                    if (this is XQuery) {
```

Comment:
When this starts happening it's an indication that you've put too much into your base class.  It's now so much that you need to resort to runtime type checking to figure out what kind of query it is.  I would recommend pushing this functionality down to the subclasses (NQuery vs XQuery).